client/broadcast.go

```diff
@@ -127,45 +120,76 @@ type intoAny interface {
 	AsAny() *codectypes.Any
 }
 
-// CheckTendermintError checks if the error returned from BroadcastTx is a
-// Tendermint error that is returned before the tx is submitted due to
-// precondition checks that failed. If an Tendermint error is detected, this
-// function returns the correct code back in TxResponse.
-//
-// TODO: Avoid brittle string matching in favor of error matching. This requires
-// a change to Tendermint's RPCError type to allow retrieval or matching against
-// a concrete error type.
-func CheckTendermintError(err error, tx tmtypes.Tx) *sdk.TxResponse {
-	if err == nil {
+// CheckTxBroadcastError checks if any errors occurred during BroadcastTx before the tx could be submitted
+// due to precondition checks that failed. If an error is detected, a TxResponse is returned with the appropriate
+// error code.
+func CheckTxBroadcastError(codespace string, code uint32, tx tmtypes.Tx) *sdk.TxResponse {
+	if codespace != sdkerrors.RootCodespace {
 		return nil
 	}
 
-	errStr := strings.ToLower(err.Error())
 	txHash := fmt.Sprintf("%X", tx.Hash())
 
-	switch {
-	case strings.Contains(errStr, strings.ToLower(mempool.ErrTxInCache.Error())):
+	switch code {
```

Comment:
I'm not completely following the point of this `switch`. On L127 we ensure that the codespace is `sdkerrors.RootCodespace`. I think that means each error's codespace would also be the root codespace.

Maybe this should be more like

```go
// package-level
var passthroughCodes = map[uint32]struct{}{
  sdkerrors.ErrWrongSequence.ABCICode(): {},
  // ...
}

// in function
if _, ok := passthroughCodes[code]; ok {
  return &sdk.TxResponse{
    Code: code,
    Codespace: sdkerrors.RootCodespace,
    TxHash: txHash,
  }
}

return nil
```

Or maybe the map isn't even necessary here?

```go
return &sdk.TxResponse{
  Code: code,
  Codespace: codespace,
  TxHash: txHash,
}
```

At that point the function doesn't seem necessary.

Reply:
I don't fully understand the reasoning around the choices for how error handling on broadcasting txs is performed but thinking about it, the latter method that you mention here seems like a way more reasonable approach since then you don't even need to account for every possible error. 

I'm going to try this route and do some more testing.